NDTensors/src/lib/SymmetrySectors/src/sector_product.jl

```diff
@@ -0,0 +1,255 @@
+# This files defines a structure for Cartesian product of 2 or more fusion sectors
+# e.g. U(1)×U(1), U(1)×SU2(2)×SU(3)
+
+using BlockArrays: blocklengths
+using ..LabelledNumbers: LabelledInteger, label, labelled, unlabel
+using ..GradedAxes: AbstractGradedUnitRange, GradedAxes, dual
+
+# =====================================  Definition  =======================================
+struct SectorProduct{Sectors} <: AbstractSector
+  arguments::Sectors
+  global _SectorProduct(l) = new{typeof(l)}(l)
+end
+
+SectorProduct(c::SectorProduct) = _SectorProduct(arguments(c))
+
+arguments(s::SectorProduct) = s.arguments
+
+# =================================  Sectors interface  ====================================
+function SymmetryStyle(T::Type{<:SectorProduct})
+  return arguments_symmetrystyle(arguments_type(T))
+end
+
+function quantum_dimension(::NotAbelianStyle, s::SectorProduct)
+  return mapreduce(quantum_dimension, *, arguments(s))
+end
+
+# use map instead of broadcast to support both Tuple and NamedTuple
+GradedAxes.dual(s::SectorProduct) = SectorProduct(map(dual, arguments(s)))
+
+function trivial(type::Type{<:SectorProduct})
+  return SectorProduct(arguments_trivial(arguments_type(type)))
+end
+
+# ===================================  Base interface  =====================================
+function Base.:(==)(A::SectorProduct, B::SectorProduct)
+  return arguments_isequal(arguments(A), arguments(B))
+end
+
+function Base.show(io::IO, s::SectorProduct)
+  (length(arguments(s)) < 2) && print(io, "sector")
+  print(io, "(")
+  symbol = ""
+  for p in pairs(arguments(s))
+    print(io, symbol)
+    sector_show(io, p[1], p[2])
+    symbol = " × "
+  end
+  return print(io, ")")
+end
+
+sector_show(io::IO, ::Int, v) = print(io, v)
+sector_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
+
+function Base.isless(s1::SectorProduct, s2::SectorProduct)
+  return arguments_isless(arguments(s1), arguments(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for SectorProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+arguments_type(::Type{<:SectorProduct{T}}) where {T} = T
+
+function sym_arguments_insert_unspecified(s1, s2)
+  return arguments_insert_unspecified(s1, s2), arguments_insert_unspecified(s2, s1)
+end
+
+function arguments_isequal(s1, s2)
+  return ==(sym_arguments_insert_unspecified(s1, s2)...)
+end
+
+# get clean results when mixing implementations
+arguments_isequal(nt::NamedTuple, t::Tuple) = arguments_isequal(t, nt)
+function arguments_isequal(t::Tuple, nt::NamedTuple)
+  if isempty(t)
+    return arguments_isequal((;), nt)
+  elseif isempty(nt)
+    return arguments_isequal(t, ())
+  end
+  return false
+end
+
+arguments_product(::NamedTuple{()}, t::Tuple) = t
+arguments_product(t::Tuple, ::NamedTuple{()}) = t
+arguments_product(::Tuple{}, nt::NamedTuple) = nt
+arguments_product(nt::NamedTuple, ::Tuple{}) = nt
```

Comment:
Also these could be moved into the generic `arguments_product` definition using `isempty` checks.

Reply:
Currently there is no generic `arguments_product`: it is only defined as `arguments_product(::Tuple, ::Tuple)` and `arguments_product(::NamedTuple, ::NamedTuple)`. `arguments_product(::NamedTuple, ::Tuple)` fails as not implemented (as it should), with a tolerance for the empty case.

This ends up similar to `isless`. I have a weak preference for explicitly dealing with the empty edge case and letting the generic case fail as not implemented.

Reply:
Ok there are ambiguities for `arguments_product( (), (;))` so I need a generic version.

Reply:
What about defining a generic function which tries to catch empty cases:
```julia
function arguments_product(l1, l2)
  isempty(l1) && return l2
  isempty(l2) && return l1
  return throw(ArgumentError(...))
end
```
That would then generalize to other storage types besides `Tuple` and `NamedTuple` and preclude the need for these specialized definitions.

Reply:
I'm not so worried about what particular error is thrown, I think the key is to try to make the code as compact and generic as possible, while still being nice and readable.